lib/elixir/lib/macro.ex

```diff
@@ -2618,11 +2643,38 @@ defmodule Macro do
     {[first_formatted | rest_formatted], result}
   end
 
+  defp dbg_format_ast_to_debug({:boolean_op, whole_ast, boolean_tree}, options) do
+    {_ast, final_value} = boolean_tree
+    whole_expr_code = to_string_with_colors(whole_ast, options)
+
+    formatted = [
+      [:bright, "[Boolean expression] ", :reset, whole_expr_code, ?\n],
```

Comment:
I am not sure if we should say "Boolean" because `||` and `&&` are not boolean. We could say "Logical expression" but I am not sure if that's useful. Thoughts?

Reply:
Also, keep in mind that we cannot always break the second part. For example, debugging this code should not raise:

```
true || raise "omg"
```

So we would need to reimplement those operators and stop when they stop.

Reply:
Oh yeah, that's a great catch 😮 However, reimplementing them becomes problematic if folks override with their own operators, no?

Reply:
Well I guess that's true for `|>` too, after all :think

Reply:
So maybe we don't reimplement them but instead evaluate the left side first.

    a || b || c

could be printed as:

```
a #=> ...
a || b #=> ...
a || b || c #=> ...
```

This would work as a general decomposition of left-associative operators.

WDYT?

Reply:
@josevalim in `a || b || c`, the example you gave for the dbg output would only work if `a` is falsy and `b` is also falsy, right? Otherwise we shouldn't eval `b` and `c`.

Reply:
Ah, but we don't need to, only to print it 🤔

Reply:
We print only as far as we go, which we makes senses. :)

Reply:
Great suggestions @josevalim. Took me a while to get there, but I think I have what you were originally envisioning now 🙃